src/transformed_distribution.jl

```diff
@@ -33,8 +33,9 @@ transformed(d) = transformed(d, bijector(d))
 
 Returns the constrained-to-unconstrained bijector for distribution `d`.
 """
-bijector(d::Distribution) = DistributionBijector(d)
-bijector(d::UnivariateDistribution) = TruncatedBijector(minimum(d), maximum(d))
+bijector(d::UnivariateDistribution) = Identity{0}()   # default to identity
```

Comment:
I suggest we add an assertion here that the domain is unbounded. This will avoid a silent wrong result when the domain is bounded.

Reply:
I suggest we throw as a fallback for multi and matrix variate distributions to avoid a silent wrong result.

Reply:
> I suggest we add an assertion here that the domain is unbounded. This will avoid a silent wrong result when the domain is bounded.

In the sense that we return identity iff it's unbounded, otherwise throw an error?
I think then we should just do something like
```julia
bijector(d::UnivariateDiscreteDistribution) = Identity{0}()
bijector(d::UnivariateContinuousDistribution) = TruncatedBijector(minimum(d), maximum(d))
```
and remove the impl for `TransformDistributions`. This way we'll get the expected behaviour if we have a well-defined `minimum` and `maximum` (irregardless of whether or not it's bounded), but will throw an error if this information isn't available.

> I suggest we throw as a fallback for multi and matrix variate distributions to avoid a silent wrong result.

So you're suggest _not_ implementing it for multivariate and matrix variate? And just letting a `unknown method` error be thrown? If so, I'm down for this. We just need to document this in the Turing documentation so that people are aware that if they want to create custom distributions they also need to define `bijector` for it if they're using AD-based samplers :+1:

Reply:
>So you're suggest not implementing it for multivariate and matrix variate? And just letting a unknown method error be thrown? If so, I'm down for this. We just need to document this in the Turing documentation so that people are aware that if they want to create custom distributions they also need to define bijector for it if they're using AD-based samplers 👍

I think this makes sense. We need to document how to write a custom distribution. Perhaps we can have a `domain` function that returns `Unbounded()`, `Simplex()`, `PD()`, `Positive()`, `Interval(lower, upper)`, or `Unknown`? Then if the user defines this function for the multivariate or matrix-variate distribution, we use the corresponding bijector accordingly. If `Unknown` we throw an unknown method error.

Reply:
I like that :+1: Could do some something like
```julia
domain(d::Distribution)
bijector(c::Constraint)

# Default impl for `Distribution`
bijector(d::Distribution) = bijector(domain(d))

# e.g. the following could replace `bijector(d::Exponential)`
domain(d::Exponential) = Positive()
bijector(::Positive) = Log{0}()
```

Though wouldn't it be better to _not_ have `Unknown`, and just let `domain` throw a method-not-implemented error?

Reply:
Ya sure `Unknown` is not necessary.

Reply:
So one thing that annoys me a bit is that it _seems_ like it should be possible to define a really nice interface for working with constraints/domains, at which point you could actually encode the input- and output-space into the Bijector definition itself. And you should be able to do stuff like create product-spaces from different spaces, etc. 

But I guess we have to start somewhere, and just getting something that makes it easier for the user is I guess useful enough (in constrast to having the user figure out which Bijector they need).

Reply:
Second thought; maybe we should delay the introduction of constraints to a different PR. I've done some work locally, but I run into issues where it's kind of unclear what `Bijector` to return, e.g. for a `VectorOfMultivariate`. The work here doesn't rely on introduction of constraints, so IMO we should just not do it here.

Reply:
Fair enough. Let me review it first.